community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -19,112 +19,302 @@
+                return doSynchronizedWork( batch );
+            }
+            finally
+            {
+                unlock();
+                unparkAnyWaiters();
+                markAsDone( batch );
+            }
+        }
+        return null;
+    }
+
+    private void unparkAnyWaiters()
+    {
+        WorkUnit<Material,W> nextBatch = stack.get();
+        while ( nextBatch != stackEnd )
+        {
+            if ( !nextBatch.isCancelled() )
+            {
+                nextBatch.unpark();
+                break;
+            }
+            WorkUnit<Material,W> next;
+            do
+            {
+                next = nextBatch.next;
+            }
+            while ( next == null );
+            nextBatch = next;
+        }
+    }
+
+    private void checkFailure( Throwable failure ) throws ExecutionException
+    {
+        if ( failure != null )
+        {
+            throw new ExecutionException( failure );
+        }
+    }
+
+    private boolean tryLock( int tryCount, WorkUnit<Material,W> unit, boolean otherwisePark )
+    {
+        if ( lock.compareAndSet( null, Thread.currentThread() ) )
+        {
+            // Got the lock!
+            return true;
+        }
+
+        // Did not get the lock, spend some time until our work has aither been completed,
+        // or we get the lock.
+        if ( otherwisePark )
+        {
+            if ( tryCount < 1000 )
+            {
+                // todo Thread.onSpinWait() in Java9?
+                Thread.yield();
             }
-            catch ( InterruptedException e )
+            else
             {
-                // We can't stop now, because our work has already been
-                // scheduled. So instead we're just going to reset the
-                // interruption status when we're done.
-                wasInterrupted = true;
+                unit.park( 10, TimeUnit.MILLISECONDS );
             }
         }
-        while ( !unit.done );
+        return false;
+    }
 
-        if ( wasInterrupted )
+    private void unlock()
+    {
+        if ( lock.getAndSet( null ) != Thread.currentThread() )
         {
-            Thread.currentThread().interrupt();
+            throw new IllegalMonitorStateException(
+                    "WorkSync accidentally released a lock not owned by the current thread" );
         }
     }
 
-    private void doSynchronizedWork()
+    private Throwable doSynchronizedWork( WorkUnit<Material,W> batch )
     {
-        WorkUnit<Material,W> batch = reverse( stack.getAndSet( stackEnd ) );
         W combinedWork = combine( batch );
+        Throwable failure = null;
 
         if ( combinedWork != null )
         {
-            combinedWork.apply( material );
+            try
+            {
+                combinedWork.apply( material );
+            }
+            catch ( Throwable throwable )
+            {
+                failure = throwable;
+            }
         }
 
         markAsDone( batch );
```

Comment:
This call to `markAsDone` is now superfluous, since we call `markAsDone` in the finally-block of `tryDoWork`.